### PR TITLE
feat: cli: tenant management

### PR DESCRIPTION
**Pull request by lizzy-0323**
## Summary

Add cmds foir tenant management ：
- create, list, update, upgrade, delete
- show, scale, 
- changepwd, replaylog, switchover,  activate

## Solution Description

Associated issue #131

### CODE REVIEW

File: internal/cli/cmd/util/error.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (chris-sun-star) [inline internal/cli/cmd/util/error.go:1]: this file should not be named with error.go, the functions in this file is doing some validation, maybe name it with validator.go is a better choice

File: internal/cli/tenant/create.go [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+	}
+	if o.Timestamp != "" {
+		o.Source.Restore.Until.Timestamp = &o.Timestamp
+	}
+	return nil
+}
+
+func (o *CreateOptions) Validate() error {
+	if o.Namespace == "" {
+		return errors.New("namespace not specified")
```
Comment (chris-sun-star) [inline internal/cli/tenant/create.go:113]: namespace is not specified

**Review by chris-sun-star — COMMENTED** [review]
take care of the type and function names and log content grammar

File: internal/cli/cluster/create.go [hunk 3]
```diff
@@ -334,3 +289,67 @@ func CreateOBClusterInstance(param *CreateOptions) *v1alpha1.OBCluster {
...
+	baseFlags.StringVar(&o.Namespace, "namespace", "default", "The namespace of the cluster")
+	baseFlags.Int64Var(&o.ClusterId, "id", 0, "The id of the cluster")
+	baseFlags.StringVarP(&o.RootPassword, "root-password", "p", "", "The root password of the cluster")
+	baseFlags.StringVar(&o.Mode, "mode", "", "The mode of the cluster")
+}
+
+// AddObserverFlags adds the observer-related flags to the command.
+func (o *CreateOptions) AddObserverFlags(cmd *cobra.Command) {
+	observerFlags := pflag.NewFlagSet("observer", pflag.ContinueOnError)
+	observerFlags.StringVar(&o.OBServer.Image, "image", "oceanbase/oceanbase-cloud-native:4.2.1.6-106000012024042515", "The image of the observer")
```
[the remark this answers is not in the document]
Comment (lizzy-0323) [reply internal/cli/cluster/create.go:321]: Thanks for your suggestion, and i will move the default values to `enter.go`

File: internal/cli/generic/generic.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"errors"
+
+	"github.com/spf13/cobra"
+
+	"github.com/oceanbase/ob-operator/api/v1alpha1"
+	"github.com/oceanbase/ob-operator/internal/dashboard/model/param"
+)
+
+type ResourceOptions struct {
```
[the remark this answers is not in the document]
Comment (lizzy-0323) [reply internal/cli/generic/generic.go:25]: Thank you for pointing that out.

File: internal/cli/cmd/tenant/show.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		PreRunE: o.Parse,
+		Run: func(cmd *cobra.Command, args []string) {
+			obtenant, err := clients.GetOBTenant(cmd.Context(), types.NamespacedName{
+				Namespace: o.Namespace,
+				Name:      o.Name,
+			})
+			if err != nil {
+				logger.Fatalln(err)
+			}
+			obtenantOperation, err := clients.GetOBTenantOperations(cmd.Context(), obtenant)
```
Comment (powerfooI) [inline internal/cli/cmd/tenant/show.go:45]: Had better name a variable containing many entities to plural format or `xxxList` to get more semantical meaning.

File: internal/cli/cmd/tenant/update.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+// NewUpdateCmd update an obtenant
+func NewUpdateCmd() *cobra.Command {
+	o := tenant.NewUpdateOptions()
+	logger := cmdUtil.GetDefaultLoggerInstance()
+	cmd := &cobra.Command{
+		Use:     "update <tenant_name>",
+		Short:   "Update ob tenant",
+		Long:    "Update ob tenant, support unitNumber/charset/connectWhiteList/priority of zones",
```
Comment (powerfooI) [inline internal/cli/cmd/tenant/update.go:32]: `charset` has already been removed from the valid operations. The description need to be updated.

File: internal/cli/tenant/activate.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func NewActivateOptions() *ActivateOptions {
+	return &ActivateOptions{}
+}
+
+func GetActivateOperation(o *ActivateOptions) *v1alpha1.OBTenantOperation {
+	activateOp := &v1alpha1.OBTenantOperation{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      o.Name + "-change-role-" + rand.String(6),
```
Comment (powerfooI) [inline internal/cli/tenant/activate.go:39]: Naming it `Name-activate-<random>` is OK.

File: internal/cli/tenant/switchover.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+func (o *SwitchOverOptions) Parse(_ *cobra.Command, args []string) error {
+	o.PrimaryTenant = args[0]
+	o.StandbyTenant = args[1]
+	return nil
+}
+
+func GetSwitchOverOperation(o *SwitchOverOptions) *v1alpha1.OBTenantOperation {
+	switchOverOp := &v1alpha1.OBTenantOperation{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      o.PrimaryTenant + "-change-role-" + rand.String(6),
```
Comment (powerfooI) [inline internal/cli/tenant/switchover.go:47]: The same as `activate`, naming it to `xx-switchover-<random>` is more precise.

File: internal/cli/utils/utils.go [hunk 9]
```diff
@@ -45,27 +48,28 @@ func GenerateUserSecrets(clusterName string, clusterId int64) *apitypes.OBUserSe
...
 	}
-	return GenerateClusterId()
+	return GenerateClusterID()
 }
 
+// CheckResourceName checks resource name in k8s
 func CheckResourceName(name string) bool {
 	regex := `[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*`
 
 	re, err := regexp.Compile(regex)
```
Comment (powerfooI) [inline internal/cli/utils/utils.go:64]: Use `regexp.MustCompile` to save the following lines.

File: internal/cli/utils/utils.go [hunk 10]
```diff
@@ -97,10 +101,22 @@ func CheckPassword(password string) bool {
 	return countUppercase >= 2 && countLowercase >= 2 && countNumber >= 2 && countSpecialChar >= 2
 }
 
+// CheckTenantName check Tenant name when creating tenant
+func CheckTenantName(name string) bool {
+	regex := `^[_a-zA-Z][^-]*$`
+
+	re, err := regexp.Compile(regex)
```
Comment (powerfooI) [inline internal/cli/utils/utils.go:108]: The same as the above situation

File: internal/cli/cmd/tenant/changepwd.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+			if err != nil {
+				logger.Fatalln(err)
+			}
+			if err := cmdUtil.CheckTenantStatus(obtenant); err != nil {
+				logger.Fatalln(err)
+			}
+			if err := tenant.GenerateNewPwd(cmd.Context(), o); err != nil {
+				logger.Fatalln(err)
+			} else {
+				logger.Println("New password generated success")
```
[the remark this answers is not in the document]
Comment (lizzy-0323) [reply internal/cli/cmd/tenant/changepwd.go:51]: Thanks for your suggestions, i will take care of the grammar.

**Review by powerfooI — APPROVED** [review]
LGTM

**Review by chris-sun-star — APPROVED** [review]
Merge this PR first, I actually mean to define the default values in a seperate file, such as "local-path", you may modify these in next PR

**Pull request merged**